src/csg_tree.cpp

```diff
@@ -566,76 +444,188 @@ void CsgOpNode::BatchUnion() const {
     // child should be quite complicated
-    std::swap(children_.front(), children_.back());
+    std::swap(children.front(), children.back());
   }
 }
 
-/**
- * Flatten the children to a list of leaf nodes and return them.
- * If forceToLeafNodes is true, the list will be guaranteed to be a list of leaf
- * nodes (i.e. no ops). Otherwise, the list may contain ops. Note that this
- * function will not apply the transform to children, as they may be shared with
- * other nodes.
- */
-std::vector<std::shared_ptr<CsgNode>> &CsgOpNode::GetChildren(
-    bool forceToLeafNodes) const {
+CsgOpNode::CsgOpNode() {}
+
+CsgOpNode::CsgOpNode(const std::vector<std::shared_ptr<CsgNode>> &children,
+                     OpType op)
+    : impl_(Impl{}), op_(op) {
   auto impl = impl_.GetGuard();
+  impl->children_ = children;
+}
 
-  if (forceToLeafNodes && !impl->forcedToLeafNodes_) {
-    impl->forcedToLeafNodes_ = true;
-    for_each(ExecutionPolicy::Par, impl->children_.begin(),
-             impl->children_.end(), [](auto &child) {
-               if (child->GetNodeType() != CsgNodeType::Leaf) {
-                 child = child->ToLeafNode();
-               }
-             });
-  }
-  return impl->children_;
+CsgOpNode::CsgOpNode(std::vector<std::shared_ptr<CsgNode>> &&children,
+                     OpType op)
+    : impl_(Impl{}), op_(op) {
+  auto impl = impl_.GetGuard();
+  impl->children_ = children;
 }
 
-void CsgOpNode::SetOp(OpType op) {
-  switch (op) {
-    case OpType::Add:
-      op_ = CsgNodeType::Union;
-      break;
-    case OpType::Subtract:
-      op_ = CsgNodeType::Difference;
-      break;
-    case OpType::Intersect:
-      op_ = CsgNodeType::Intersection;
-      break;
+std::shared_ptr<CsgNode> CsgOpNode::Boolean(
+    const std::shared_ptr<CsgNode> &second, OpType op) {
+  std::vector<std::shared_ptr<CsgNode>> children;
+  children.push_back(shared_from_this());
+  children.push_back(second);
+
+  return std::make_shared<CsgOpNode>(children, op);
+}
+
+std::shared_ptr<CsgNode> CsgOpNode::Transform(const mat3x4 &m) const {
+  auto node = std::make_shared<CsgOpNode>();
+  node->impl_ = impl_;
+  node->transform_ = m * Mat4(transform_);
+  node->op_ = op_;
+  return node;
+}
+
+struct CsgStackFrame {
+  bool finalize;
+  OpType parent_op;
+  mat3x4 transform;
+  std::vector<std::shared_ptr<CsgLeafNode>> *parent_children;
+  std::shared_ptr<const CsgOpNode> op_node;
+  std::vector<std::shared_ptr<CsgLeafNode>> left_children;
+  std::vector<std::shared_ptr<CsgLeafNode>> right_children;
+
+  CsgStackFrame(bool finalize, OpType parent_op, mat3x4 transform,
+                std::vector<std::shared_ptr<CsgLeafNode>> *parent_children,
+                std::shared_ptr<const CsgOpNode> op_node)
+      : finalize(finalize),
+        parent_op(parent_op),
+        transform(transform),
+        parent_children(parent_children),
+        op_node(op_node) {}
+};
+
+std::shared_ptr<CsgLeafNode> CsgOpNode::ToLeafNode() const {
```

Comment:
I never fully understood what the previous version of this algorithm did either - perhaps some comments on the general steps this is recursing through?